internal/rest/avi_evh_obj.go

```diff
@@ -0,0 +1,805 @@
+	pathMatches = append(pathMatches, &path_match)
+	vhMatch := &avimodels.VHMatch{Host: &vs_meta.EvhHostName, Path: pathMatches}
+	vhMatches := make([]*avimodels.VHMatch, 0)
+	vhMatches = append(vhMatches, vhMatch)
+	evhChild.VhMatches = vhMatches
+
+	if vs_meta.DefaultPool != "" {
+		pool_ref := "/api/pool/?name=" + vs_meta.DefaultPool
+		evhChild.PoolRef = &pool_ref
+	}
+	if len(vs_meta.PoolGroupRefs) > 0 {
+		poolgroup_ref := "/api/poolgroup/?name=" + vs_meta.PoolGroupRefs[0].Name
+		evhChild.PoolGroupRef = &poolgroup_ref
+	}
+
+	// No need of HTTP rules for TLS passthrough.
+	if vs_meta.TLSType != utils.TLS_PASSTHROUGH {
+		// this overwrites the sslkeycert created from the Secret object, with the one mentioned in HostRule.TLS
+		if vs_meta.SSLKeyCertAviRef != "" {
+			evhChild.SslKeyAndCertificateRefs = append(evhChild.SslKeyAndCertificateRefs, vs_meta.SSLKeyCertAviRef)
+		} else {
+			for _, sslkeycert := range vs_meta.SSLKeyCertRefs {
+				certName := "/api/sslkeyandcertificate/?name=" + sslkeycert.Name
+				evhChild.SslKeyAndCertificateRefs = append(evhChild.SslKeyAndCertificateRefs, certName)
+			}
+		}
+
+		var httpPolicyCollection []*avimodels.HTTPPolicies
+		internalPolicyIndexBuffer := int32(11)
+		for i, http := range vs_meta.HttpPolicyRefs {
+			// Update them on the VS object
+			var j int32
+			j = int32(i) + internalPolicyIndexBuffer
+			httpPolicy := fmt.Sprintf("/api/httppolicyset/?name=%s", http.Name)
+			httpPolicies := &avimodels.HTTPPolicies{HTTPPolicySetRef: &httpPolicy, Index: &j}
+			httpPolicyCollection = append(httpPolicyCollection, httpPolicies)
+		}
+
+		// from hostrule CRD
+		bufferLen := int32(len(httpPolicyCollection)) + internalPolicyIndexBuffer + 5
+		for i, policy := range vs_meta.HttpPolicySetRefs {
+			var j int32
+			j = int32(i) + bufferLen
+			httpPolicy := policy
+			httpPolicies := &avimodels.HTTPPolicies{HTTPPolicySetRef: &httpPolicy, Index: &j}
+			httpPolicyCollection = append(httpPolicyCollection, httpPolicies)
+		}
+
+		evhChild.HTTPPolicies = httpPolicyCollection
+	}
+
+	var rest_ops []*utils.RestOp
+	var rest_op utils.RestOp
+	var path string
+	if rest_method == utils.RestPut {
+
+		path = "/api/virtualservice/" + cache_obj.Uuid
+		rest_op = utils.RestOp{Path: path, Method: rest_method, Obj: evhChild,
+			Tenant: vs_meta.Tenant, Model: "VirtualService", Version: utils.CtrlVersion}
+		rest_ops = append(rest_ops, &rest_op)
+
+	} else {
+
+		macro := utils.AviRestObjMacro{ModelName: "VirtualService", Data: evhChild}
+		path = "/api/macro"
+		rest_op = utils.RestOp{Path: path, Method: rest_method, Obj: macro,
+			Tenant: vs_meta.Tenant, Model: "VirtualService", Version: utils.CtrlVersion}
+		rest_ops = append(rest_ops, &rest_op)
+
+	}
+
+	return rest_ops
+}
+
+func (rest *RestOperations) ExecuteRestAndPopulateCacheForEvh(rest_ops []*utils.RestOp, aviObjKey avicache.NamespaceName, avimodel *nodes.AviObjectGraph, key string, sslKey ...utils.NamespaceName) bool {
```

Comment:
Can we reuse ExecuteRestAndPopulateCache and RefreshCacheForRetryLayerForEvh functions ?

Reply:
Addressed this by introducing the interface.